File: api/services/export-tiktok.js
import fs from 'fs'

import AdmZip from 'adm-zip'

import {

  XMLParser,

  XMLBuilder

} from 'fast-xml-parser'

export async function exportTiktok({

  data,
  templatePath,
  outputPath

}) {

  // =========================
  // COPY TEMPLATE
  // =========================

  fs.copyFileSync(
    templatePath,
    outputPath
  )

  // =========================
  // OPEN ZIP
  // =========================

  const zip =
    new AdmZip(
      outputPath
    )

  // =========================
  // XML TOOLS
  // =========================

  const parser =
    new XMLParser({

      ignoreAttributes:
        false
    })

  const builder =
    new XMLBuilder({

      ignoreAttributes:
        false
    })

  // =========================
  // LOAD SHEET XML
  // =========================

  const sheetEntry =
    zip.getEntry(
      'xl/worksheets/sheet1.xml'
    )

  if (!sheetEntry) {

    throw new Error(
      'sheet1.xml not found'
    )
  }

  const sheetXml =
    sheetEntry
      .getData()
      .toString('utf8')

  const sheetObj =
    parser.parse(
      sheetXml
    )

  const rows =
    sheetObj
      .worksheet
      .sheetData
      .row

  let updated = 0

  // =========================
  // PROCESS ROWS
  // =========================

  for (const row of rows) {

    const cells =
      Array.isArray(row.c)
        ? row.c
        : [row.c]

    // =========================
    // D = SKU ID
    // G = QUANTITY
    // =========================

    const skuIdCell =
      cells.find(cell =>

        cell['@_r']
          ?.startsWith(
            'D'
          )
      )

    const qtyCell =
      cells.find(cell =>

        cell['@_r']
          ?.startsWith(
            'G'
          )
      )

    if (
      !skuIdCell ||
      !qtyCell
    ) {

      continue
    }

    // =========================
    // GET SKU ID
    // =========================

    const skuId =
      String(
        skuIdCell.v || ''
      )
        .replace(/\.0$/, '')
        .replace(/\s/g, '')
        .trim()

    // SKIP HEADER

    if (
      !skuId ||
      skuId === 'SKU ID'
    ) {

      continue
    }

    // =========================
    // FIND PRODUCT
    // =========================

    const product =
      data.find(item => {

        const dbSku =
          String(
            item.tiktok_sku_id || ''
          )
            .replace(/\.0$/, '')
            .replace(/\s/g, '')
            .trim()

        return (
          dbSku === skuId
        )
      })

    if (!product) {

      console.log(
        'NOT FOUND:',
        skuId
      )

      continue
    }

    // =========================
    // UPDATE QUANTITY ONLY
    // =========================

    qtyCell.v =
      Number(
        product.stock || 0
      )

    delete qtyCell['@_t']
    delete qtyCell.is

    updated++

    console.log(
      'UPDATED:',
      skuId,
      product.stock
    )
  }

  // =========================
  // BUILD XML
  // =========================

  const newSheetXml =
    builder.build(
      sheetObj
    )

  // =========================
  // UPDATE ZIP
  // =========================

  zip.updateFile(

    'xl/worksheets/sheet1.xml',

    Buffer.from(
      newSheetXml
    )
  )

  // =========================
  // SAVE ZIP
  // =========================

  zip.writeZip(
    outputPath
  )

  return {

    updated,

    outputPath
  }
}

File: api/python/export_tiktok.py
import sys
import json

from openpyxl import load_workbook


# =========================
# ARGUMENTS
# =========================

template_path = sys.argv[1]
output_path = sys.argv[2]
json_data = sys.argv[3]

products = json.loads(json_data)


# =========================
# LOAD WORKBOOK
# =========================

wb = load_workbook(template_path)

ws = wb['Template']


# =========================
# FIND HEADER ROW
# =========================

header_row = None

for row in ws.iter_rows():

    values = [
        str(cell.value).strip()
        if cell.value is not None
        else ''
        for cell in row
    ]

    if (
        'SKU ID' in values
        and
        'Quantity' in values
    ):

        header_row = row[0].row
        break


if not header_row:

    raise Exception(
        'HEADER NOT FOUND'
    )


# =========================
# FIND COLUMN
# =========================

sku_col = None
qty_col = None

for cell in ws[header_row]:

    value = (
        str(cell.value)
        .strip()
        .lower()
    )

    if value == 'sku id':

        sku_col = cell.column

    if value == 'quantity':

        qty_col = cell.column


if not sku_col or not qty_col:

    raise Exception(
        'COLUMN NOT FOUND'
    )


# =========================
# UPDATE STOCK
# =========================

updated = 0

for row in ws.iter_rows(
    min_row=header_row + 1
):

    sku_id = row[
        sku_col - 1
    ].value

    if not sku_id:
        continue

    sku_id = (
        str(sku_id)
        .replace('.0', '')
        .replace(' ', '')
        .strip()
    )

    product = next(

        (
            item for item in products

            if
            str(
                item.get(
                    'tiktok_sku_id',
                    ''
                )
            )
            .replace('.0', '')
            .replace(' ', '')
            .strip()

            == sku_id
        ),

        None
    )

    if not product:
        continue

    # =========================
    # UPDATE QUANTITY
    # =========================

    row[
        qty_col - 1
    ].value = int(
        product.get(
            'stock',
            0
        )
    )

    updated += 1


# =========================
# SAVE
# =========================

wb.save(output_path)

print(
    json.dumps({

        'updated': updated,

        'outputPath':
            output_path
    })
)